openglyph/src/io/mega_file.cpp

```diff
@@ -0,0 +1,169 @@
+#include <khepri/utility/crc32.hpp>
+#include <khepri/utility/string.hpp>
+
+#include <openglyph/io/mega_file.hpp>
+namespace openglyph::io {
+
+class MegaFile::SubFile : public khepri::io::Stream
+{
+public:
+    SubFile(const SubFileInfo info, khepri::io::File* file);
+
+    SubFile(const SubFile&)            = delete;
+    SubFile(SubFile&&)                 = delete;
+    SubFile& operator=(const SubFile&) = delete;
+    SubFile& operator=(SubFile&&)      = delete;
+
+    /// \see stream::readable
+    [[nodiscard]] bool readable() const noexcept override
+    {
+        return true;
+    }
+
+    /// \see stream::writable
+    [[nodiscard]] bool writable() const noexcept override
+    {
+        return false;
+    }
+
+    /// \see stream::seekable
+    [[nodiscard]] bool seekable() const noexcept override
+    {
+        return true;
+    }
+
+    /**
+     * \brief Reads data from the stream at the current position.
+     *
+     * \param[in] buffer the memory to write the read data to.
+     * \param[in] count the number of consecutive bytes to read.
+     *
+     * \return the number of bytes read and stored in \a buffer.
+     */
+    virtual size_t read(void* buffer, size_t count) override;
+
+    /**
+     * \brief Writes data to the stream at the current position (not supported for subfiles).
+     *
+     * \param[in] buffer pointer to the data to write.
+     * \param[in] count the number of consecutive bytes to write.
+     *
+     * \return the number of bytes from \a buffer written to the stream.
+     */
+    virtual size_t write(const void* buffer, size_t count) override;
+
+    /**
+     * \brief Changes the file position
+     *
+     * \param[in] offset offset by how much to change the file position
+     * \param[in] origin the origin from whence to change the position
+     *
+     * \return the new file position, from the start of the file
+     */
+    virtual long long seek(long long offset, khepri::io::SeekOrigin origin) override;
+
+private:
+    const SubFileInfo info;
+    khepri::io::File* mega_file;
+    std::uint64_t     local_read_offset;
+};
+
+MegaFile::MegaFile(const std::filesystem::path& mega_file_path)
+    : m_file(std::make_unique<khepri::io::File>(mega_file_path, khepri::io::OpenMode::read))
+{
+    std::tie(m_filenames, m_fileinfo) = extract_metadata();
+
+    // calculate hashes for this session
```

Comment:
Why, `m_fileinfo` already has hashes? Why are you copying those into a separate vector?